<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Pandas:-cargando-archivos-de-datos" data-toc-modified-id="Pandas:-cargando-archivos-de-datos-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Pandas: cargando archivos de datos</a></span><ul class="toc-item"><li><span><a href="#pd.read_csv" data-toc-modified-id="pd.read_csv-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>pd.read_csv</a></span></li><li><span><a href="#pd.read_excel" data-toc-modified-id="pd.read_excel-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>pd.read_excel</a></span></li></ul></li><li><span><a href="#Pandas,-guardando-datos" data-toc-modified-id="Pandas,-guardando-datos-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Pandas, guardando datos</a></span><ul class="toc-item"><li><span><a href="#Guardar-en-formato-csv" data-toc-modified-id="Guardar-en-formato-csv-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Guardar en formato csv</a></span></li><li><span><a href="#DataFrame.to_csv" data-toc-modified-id="DataFrame.to_csv-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>DataFrame.to_csv</a></span></li><li><span><a href="#Guardar-en-formato-Excel" data-toc-modified-id="Guardar-en-formato-Excel-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Guardar en formato Excel</a></span></li><li><span><a href="#DataFrame.to_excel" data-toc-modified-id="DataFrame.to_excel-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>DataFrame.to_excel</a></span></li></ul></li><li><span><a href="#Nombres-de-las-columnas:---.columns" data-toc-modified-id="Nombres-de-las-columnas:---.columns-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Nombres de las columnas:   .columns</a></span></li><li><span><a href="#INSPECCION-DEL-DATAFRAME" data-toc-modified-id="INSPECCION-DEL-DATAFRAME-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>INSPECCION DEL DATAFRAME</a></span><ul class="toc-item"><li><span><a href="#Tipos-de-Datos" data-toc-modified-id="Tipos-de-Datos-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Tipos de Datos</a></span></li></ul></li><li><span><a href="#ACCESO-A-COLUMNAS" data-toc-modified-id="ACCESO-A-COLUMNAS-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>ACCESO A COLUMNAS</a></span><ul class="toc-item"><li><span><a href="#Por-nombre-con-[--]" data-toc-modified-id="Por-nombre-con-[--]-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Por nombre con [  ]</a></span></li><li><span><a href="#Por-nombre-con-punto-(como-atributo)" data-toc-modified-id="Por-nombre-con-punto-(como-atributo)-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Por nombre con punto (como atributo)</a></span></li><li><span><a href="#Por-posición---.iloc" data-toc-modified-id="Por-posición---.iloc-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Por posición   .iloc</a></span></li><li><span><a href="#Por-nombre-con-.loc" data-toc-modified-id="Por-nombre-con-.loc-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Por nombre con .loc</a></span></li></ul></li><li><span><a href="#ACCESO-A-FILAS" data-toc-modified-id="ACCESO-A-FILAS-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>ACCESO A FILAS</a></span></li><li><span><a href="#EJERCICIO" data-toc-modified-id="EJERCICIO-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>EJERCICIO</a></span></li><li><span><a href="#Crear-un-Índice-para-las-filas" data-toc-modified-id="Crear-un-Índice-para-las-filas-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Crear un Índice para las filas</a></span><ul class="toc-item"><li><span><a href="#.unique" data-toc-modified-id=".unique-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>.unique</a></span></li><li><span><a href="#Ejercicio" data-toc-modified-id="Ejercicio-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Ejercicio</a></span></li><li><span><a href="#EJERCICIO" data-toc-modified-id="EJERCICIO-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>EJERCICIO</a></span></li><li><span><a href="#EJERCICIO" data-toc-modified-id="EJERCICIO-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>EJERCICIO</a></span></li></ul></li></ul></div>

![IES21](img/logo_ies.png)

# COMPUTACIÓN CIENTÍFICA
# Situación Profesional 2
# CLASE 2: Pandas
**Prof. Ricardo Piña**

In [1]:
import pandas as pd
import numpy as np

## Pandas: cargando archivos de datos

La mayoría de las veces deberemos acceder a datos que fueron obtenidos de alguna otra fuente, como habíamos comentado, Pandas capaz de leer y guardar datos en diversos formatos como: 
    - csv, archivos de texto, 
    - archivos de Excel
    - bases de datos SQL
    - HTML
    - XML
    - JSON
    - HDF5 (formato jerárquico de datos; muy veloz)

### pd.read_csv

El formato más utilizado para intercambiar archivos de dayos es el csv (Datos Separados por comas).  
En este formato lo habitual es que el **separador** de registros sea la __coma ,__ pero a veces se utilizan otros como el punto y coma ; **(típico cuando guardamos en formato csv desde Excel en español** o el **Tabulador** que se suele simbolizar como **\t**.  

Puede ver toda la documentación en https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html 


Vamos a probar con un archivo de datos que tiene ciertas "incomodidades".  
Tratemos de cargar el archivo iris.tab que tenemos en la carpeta data: 

In [2]:
df_iris=pd.read_csv("data/iris.tab")

Para ver sólo los primeros 5 registros usaremos **.head()**. También podemos ver los últimos con **.tail()**. Si desea ver otra cantidad de registros puede usar **.head(cantidad_filas)**.

In [3]:
df_iris.head()

sepal length\tsepal width\tpetal length\tpetal width\tiris
0                                      c\tc\tc\tc\td        
1                                      \t\t\t\tclass        
2                    5.1\t3.5\t1.4\t0.2\tIris-setosa        
3                    4.9\t3.0\t1.4\t0.2\tIris-setosa        
4                    4.7\t3.2\t1.3\t0.2\tIris-setosa

Parece que se ha confundido un poco.   
Veamos: 

- obviamente la primer fila de los datos tiene los **encabezados** o nombres de los campos, lo cual nos resultará de utilidad. En los casos en que éstos no están, es habitual acompañar el archivo de datos con otro archivo que hace las veces de diccionario donde se explica qué es cada columna.  
- hay dos filas que no son datos 
- los datos mismos parecen bastante confusos, como que no ha tomado bien al separador. 

Veamos qué pasa si le indicamos que el separador es el tabulador: 


In [4]:
df_iris=pd.read_csv("data/iris.tab",sep='\t')
df_iris.head()

sepal length sepal width petal length petal width         iris
0            c           c            c           c            d
1          NaN         NaN          NaN         NaN        class
2          5.1         3.5          1.4         0.2  Iris-setosa
3          4.9         3.0          1.4         0.2  Iris-setosa
4          4.7         3.2          1.3         0.2  Iris-setosa

Mejoró bastante,  

- ahora los encabezados parece que fueron bien identificados
- los datos se ven bien separados. 
- las filas con índice 1 y 2 obviamente no corresponden a datos y deberíamos **saltearlas**. 

In [5]:
df_iris=pd.read_csv("data/iris.tab",sep='\t',skiprows=(1,2))
df_iris.head()

sepal length  sepal width  petal length  petal width         iris
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Ahora sí se ve bien.  
Si bien en este caso no sería necesario, también podemos indicarle en qué fila están los encabezados, fíjese en las opciones dadas en el vínculo.  
Las opciones anteriores son las más habituales.


### pd.read_excel


Otro formato muy utilizado es el de las planillas de cálculo Excel, Pandas también es capaz de leer y guardar datos en este formato, encontrará todas las referencias en : 
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html

Veamos ahora un ejemplo:

In [5]:
df_credito=pd.read_excel('data/credito_01.xlsx')

Veamos las primeras 3 filas

In [7]:
# Escriba aquí su código
df_credito.head(3)

Tiene_Deuda Genero Trabaja Propietario Dar_Credito
0          Si      F      No          Si          No
1          Si      M      No          Si          No
2          No      F      Si          Si          Si

Veamos las últimas 4 filas

In [8]:
# Escriba aquí su código
df_credito.tail(4)

Tiene_Deuda Genero Trabaja Propietario Dar_Credito
9           Si      M      Si          No          No
10          Si      M      No          No          No
11          No      F      No          Si          No
12          No      M      Si          Si          Si

## Pandas, guardando datos

También es posible guardar los datos en los formatos estándard antes mencionados

### Guardar en formato csv

### DataFrame.to_csv

Documentación en https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html

Supongamos que queremos guardar el dataFrame df_credito en formato csv. El separador más habitual es la coma, que es el valor que tomará por defecto, pero supongamos que deseamos que sea separado por tabulación:

In [18]:
df_credito.to_csv('data/credito.tab',sep='\t')

### Guardar en formato Excel 

### DataFrame.to_excel
Documentación en: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_excel.html

Lea la documentación anterior y guarde el dataframe df_iris en formato de excel en la carperta data con el nombre iris.xlsx y en la Hoja de Excel llamada datos:

In [19]:
df_iris.to_excel('data/iris.xlsx',sheet_name='datos')

Luego revise abriendo el archivo con Excel.  
Observe que como guardamos todo el DataFrame, se guardó también la columna índice como primero columna.

## Nombres de las columnas:   .columns

Podemos acceder a los nombres de las columnas muf fácilmente:

In [20]:
df_iris.columns

Index(['sepal length\tsepal width\tpetal length\tpetal width\tiris'], dtype='object')

Como bien saben quienes programan habitualmente, no es conveniente que los nombres de los campos o columnas incluyan espacios, siempre traen problemas, así que los vamos a sustituir por nombres sin espacios.

In [21]:
df_iris.columns=['sepal_length','sepal_width','petal_length','petal width','iris'] # un errorcito a propósito
df_iris.head()

ValueError: Length mismatch: Expected axis has 1 elements, new values have 5 elements

También podríamos cambiar el nombre de una sola o algunas de las columnas

In [22]:
df_iris=df_iris.rename(columns={'petal width': 'petal_width'})
df_iris.head()

sepal length\tsepal width\tpetal length\tpetal width\tiris
0                                      c\tc\tc\tc\td        
1                                      \t\t\t\tclass        
2                    5.1\t3.5\t1.4\t0.2\tIris-setosa        
3                    4.9\t3.0\t1.4\t0.2\tIris-setosa        
4                    4.7\t3.2\t1.3\t0.2\tIris-setosa

Observe que para el cambio realmente se produzca en el df_iris, tuvimos que asignarle el resultado de rename.

In [23]:
df_iris.columns

Index(['sepal length\tsepal width\tpetal length\tpetal width\tiris'], dtype='object')

## INSPECCION DEL DATAFRAME

Generalmente luego de cargar un DataFrame queremos ver cierta información referida a él. Podemos obtener un buen vistazo usando:  

.info()  
.describe()

In [24]:
df_iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 1 columns):
 #   Column                                                  Non-Null Count  Dtype 
---  ------                                                  --------------  ----- 
 0   sepal length	sepal width	petal length	petal width	iris  152 non-null    object
dtypes: object(1)
memory usage: 1.3+ KB


### Tipos de Datos

Conocer los tipos de datos que conforman cada columna suele ser de gran utilidad, lamentablemente los tipos de datos de Pandas no son exactamente los mismos que en Python:  

|   Pandas   |  Python  |                    Descripción                   |
|:----------:|:--------:|:------------------------------------------------:|
| object     | string   | El tipo de datos más común. Cadena de caracteres |
| int64      | int      | Números enteros                                  |
| float64    | float    | Números decimales                                |
| datetime64 | datetime | Fecha y Hora                                     |



Le podemos pedir a Pandas que nos **describa** el DataFrame. 

In [16]:
# Escriba aquí su código


sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

En este caso nos devuelve un buen resumen de información de **estadística descriptiva** sobre los valores contenidos en cada una de las columnas.  

**Nota** Por defecto sólo devuelve esfa información para las columnas cuyos valores son numéricos, si deseamos información sobre columnas tipo objetct devemos utilizar  

**.describe(include=all)**

También podemos usar .info()

In [1]:
df_iris.info()

NameError: name 'df_iris' is not defined

## ACCESO A COLUMNAS

En Pandas existen diversas maneras de acceder a las columnas:

- por nombre

- por posición


### Por nombre con [  ]

Recordemos las columnas de df_iris

In [18]:
df_iris.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'iris'], dtype='object')

Supongamos que necesitamos acceder sólo a la columna 'sepal_width'.  

Veamos las primeras 5 filas de esta columna

In [19]:
df_iris['sepal_width'].head()

0    3.5
1    3.0
2    3.2
3    3.1
4    3.6
Name: sepal_width, dtype: float64

Supongamos que necesitamos acceder a la columna 'sepal_width' y a la columna iris: 

In [20]:
df_iris[['sepal_width','iris']].head() # observe el doble corchete

sepal_width         iris
0          3.5  Iris-setosa
1          3.0  Iris-setosa
2          3.2  Iris-setosa
3          3.1  Iris-setosa
4          3.6  Iris-setosa

### Por nombre con punto (como atributo)

También es posible acceder a la fila sepal_width de la siguiente manera:

In [21]:
df_iris.sepal_width.head()

0    3.5
1    3.0
2    3.2
3    3.1
4    3.6
Name: sepal_width, dtype: float64

### Por posición   .iloc

Para acceder a las filas por su ubicación nos debemos valer del índice que tienen las columnas (recuerde que Python indexa desde 0).  

Esta forma de acceder es muy parecida a la de Numpy con  

**.iloc[núm fila,núm columna]**,   

excepto que debe ir precedida de **.iloc**.

Para ver la primer columna indicando su índice:

In [22]:
df_iris.iloc[:,0].head() # todas las filas, la columna 0 es decir la primera

0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
Name: sepal_length, dtype: float64

Muestre últimos elementos de la columna de índice 3

In [23]:
# Escriba aquí su código


145    2.3
146    1.9
147    2.0
148    2.3
149    1.8
Name: petal_width, dtype: float64

Muestre las primeras filas de las 3 primeras columnas:

In [24]:
df_iris.iloc[:,0:3].head()   # todas las filas, las columnas de 0 a 3 sin incluir la de índice 3

sepal_length  sepal_width  petal_length
0           5.1          3.5           1.4
1           4.9          3.0           1.4
2           4.7          3.2           1.3
3           4.6          3.1           1.5
4           5.0          3.6           1.4

Muestre todas las columnas de índice par.  


Primero hagamos una prueba sencilla:


Como sabemos que las columnas de índice para son la 0, 2 y 4, podríamos hacer:

In [25]:
df_iris.iloc[:,[0,2,4]].head()

sepal_length  petal_length         iris
0           5.1           1.4  Iris-setosa
1           4.9           1.4  Iris-setosa
2           4.7           1.3  Iris-setosa
3           4.6           1.5  Iris-setosa
4           5.0           1.4  Iris-setosa

Donde vemos las columnas solicitadas.  

Pero queremos que Ud lo haga de manera que se adapte a *cualquier* DataFrame, sin importar cuántas columnas tuviera ... (puede crear un rango para las columnas pares como hicimos en las primeras clases ... si necesita saber cúantas columnas tiene si DataFrame ... puede probar con shape ...)


In [26]:
# Escriba aquí su código


sepal_length  petal_length         iris
0           5.1           1.4  Iris-setosa
1           4.9           1.4  Iris-setosa
2           4.7           1.3  Iris-setosa
3           4.6           1.5  Iris-setosa
4           5.0           1.4  Iris-setosa

### Por nombre con .loc

Accedamos a la columna sepal_length

In [27]:
df_iris.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'iris'], dtype='object')

In [28]:
df_iris.loc[:, 'sepal_length'].head()

0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
Name: sepal_length, dtype: float64

Algo muy interesante que tiene esta nomenclatura es que nos permite utilizar rangos ... pero con los nombres.

Qué piensa que hará:  

df_iris.loc[:,'sepal_length':'petal_length']

In [29]:
df_iris.loc[:,'sepal_length':'petal_length'].head()

sepal_length  sepal_width  petal_length
0           5.1          3.5           1.4
1           4.9          3.0           1.4
2           4.7          3.2           1.3
3           4.6          3.1           1.5
4           5.0          3.6           1.4

## ACCESO A FILAS

Miremos el df_iris

In [30]:
df_iris.head()

sepal_length  sepal_width  petal_length  petal_width         iris
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Observemos los números de la izquierda, si bien están encolumnados, no son una columna de datos, sino que son los índices de las filas.  
Puede pensar en ellos como si fueran los nombres de las filas, asi como antes teníamos los nombres de las columnas. 

Por defecto, Pandas, indexa con números enteros no negativos, pero si fuera necesario, se podría cambiar.  

Un caso típico donde se cambian los índices por defecto por otros es en el caso de las **Series de Tiempo**, donde se suele tomar como índice las Fechas en las cuales se midieron los datos de la serie.   


Veamos la primer fila, es decir la de índice 0 usando loc

In [31]:
df_iris.loc[0]    #podríamos haber usado[0,] o [o,:] y hubiéramos obtenido los mismos resultados.

sepal_length            5.1
sepal_width             3.5
petal_length            1.4
petal_width             0.2
iris            Iris-setosa
Name: 0, dtype: object

Muestre la centésima fila:

In [32]:
df_iris.loc[99]

sepal_length                5.7
sepal_width                 2.8
petal_length                4.1
petal_width                 1.3
iris            Iris-versicolor
Name: 99, dtype: object

Muestre la última fila con loc, considerando que su código debe funcionar cualquiera sea la cantidad de filas que tenga el DataFrame.

In [33]:
# Escriba aquí su código


sepal_length               5.9
sepal_width                  3
petal_length               5.1
petal_width                1.8
iris            Iris-virginica
Name: 149, dtype: object

Ahora también muestre la última fila, pero esta vez use tail

In [34]:
# Escriba aquí su código


sepal_length  sepal_width  petal_length  petal_width            iris
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

... más cómodo, no?

Habrá notado que los resultados se muestran de distinta manera, es porque ambos son distintos tipos de objetos.  

Utilice **type( )** para que Python le indique  qué es cada uno de ellos.

Primero con .loc

In [35]:
# Escriba aquí su código
type(df_iris.loc[df_iris.shape[0]-1])

pandas.core.series.Series

Podemos ver que es una Serie

Ahora con tail

In [36]:
# Escriba aquí su código


pandas.core.frame.DataFrame

Es un DataFrame. 

También podemos acceder con iloc a varias filas, de forma similar a como lo hacíamos con las columnas.  

Obtenga las filas de índice 10, 50 y 100

In [37]:
# Escriba aquí su código

df_iris.iloc[[10,50,100]]

sepal_length  sepal_width  petal_length  petal_width             iris
10            5.4          3.7           1.5          0.2      Iris-setosa
50            7.0          3.2           4.7          1.4  Iris-versicolor
100           6.3          3.3           6.0          2.5   Iris-virginica

En este caso como los índices son los números, no habría diferencia entre usar .iloc o .loc.

## EJERCICIO

Cargue el archivo datos_meses.xlsx en el DataFrame df_meses, que corresponde a las hipotéticas facturaciones de una empresa.
Luego vea todo el dataframe.

In [45]:
#EScriba aquí su código
df_meses=pd.read_excel("data/datos_meses.xlsx")
df_meses

Mes  2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  \
0        Enero   100   120   135   125   245   275   170   220   100   280   
1      Febrero   120   125   150   180   200   220   150   200   120   280   
2        Marzo   135   180   180   125   170   200   120   150    25   275   
3        Abril   150   125   250   280   150   150   105   100   120   220   
4         Mayo   180   280   245   280   120   135   100   250   280   200   
5        Junio   250   280   200   275   105   120   120   150    25   150   
6        Julio   245   275   170   220   100   150   125   250   280   100   
7       Agosto   200   220   150   200   120   180   280   245   280   120   
8   Septiembre   170   200   120   150    25   250   280   150   150   135   
9      Octubre   150   150   105   100   120   135   125   245   100   150   
10   Noviembre   120   135   100   250   280   200   275   105   250   180   
11   Diciembre   105   120   120   150    25   250   280   150   150   250   

    2020  
0    150  
1    120  
2    105  
3    100  
4    120  
5     25  
6    120  
7    125  
8    180  
9    125  
10   280  
11   280

En este caso estaría espectacular indexar las filas con los nombres de los meses ...

## Crear un Índice para las filas

In [46]:
df_meses.set_index('Mes')

2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  2020
Mes                                                                         
Enero        100   120   135   125   245   275   170   220   100   280   150
Febrero      120   125   150   180   200   220   150   200   120   280   120
Marzo        135   180   180   125   170   200   120   150    25   275   105
Abril        150   125   250   280   150   150   105   100   120   220   100
Mayo         180   280   245   280   120   135   100   250   280   200   120
Junio        250   280   200   275   105   120   120   150    25   150    25
Julio        245   275   170   220   100   150   125   250   280   100   120
Agosto       200   220   150   200   120   180   280   245   280   120   125
Septiembre   170   200   120   150    25   250   280   150   150   135   180
Octubre      150   150   105   100   120   135   125   245   100   150   125
Noviembre    120   135   100   250   280   200   275   105   250   180   280
Diciembre    105   120   120   150    25   250   280   150   150   250   280

Fíjese que ha creado el índice con los meses, y que ya no lo presenta como una columna, observe que la palabra mes está más abajo.

Sin embargo, fíjese qué pasa al mostrar el dataframe

In [47]:
df_meses

Mes  2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  \
0        Enero   100   120   135   125   245   275   170   220   100   280   
1      Febrero   120   125   150   180   200   220   150   200   120   280   
2        Marzo   135   180   180   125   170   200   120   150    25   275   
3        Abril   150   125   250   280   150   150   105   100   120   220   
4         Mayo   180   280   245   280   120   135   100   250   280   200   
5        Junio   250   280   200   275   105   120   120   150    25   150   
6        Julio   245   275   170   220   100   150   125   250   280   100   
7       Agosto   200   220   150   200   120   180   280   245   280   120   
8   Septiembre   170   200   120   150    25   250   280   150   150   135   
9      Octubre   150   150   105   100   120   135   125   245   100   150   
10   Noviembre   120   135   100   250   280   200   275   105   250   180   
11   Diciembre   105   120   120   150    25   250   280   150   150   250   

    2020  
0    150  
1    120  
2    105  
3    100  
4    120  
5     25  
6    120  
7    125  
8    180  
9    125  
10   280  
11   280

Vuelve a mostrar a Mes como una simple columna.  
Esto es así porque debemos asignar el índice al dataframe:

In [48]:
df_meses=df_meses.set_index('Mes')
df_meses

2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  2020
Mes                                                                         
Enero        100   120   135   125   245   275   170   220   100   280   150
Febrero      120   125   150   180   200   220   150   200   120   280   120
Marzo        135   180   180   125   170   200   120   150    25   275   105
Abril        150   125   250   280   150   150   105   100   120   220   100
Mayo         180   280   245   280   120   135   100   250   280   200   120
Junio        250   280   200   275   105   120   120   150    25   150    25
Julio        245   275   170   220   100   150   125   250   280   100   120
Agosto       200   220   150   200   120   180   280   245   280   120   125
Septiembre   170   200   120   150    25   250   280   150   150   135   180
Octubre      150   150   105   100   120   135   125   245   100   150   125
Noviembre    120   135   100   250   280   200   275   105   250   180   280
Diciembre    105   120   120   150    25   250   280   150   150   250   280

Ahora sí, tenemos un índice aplicado a nuestro dataframe, y podemos usarlo.

Muestre la factuaración del mes de Enero de todos los años con la nomenclatura .loc

In [39]:
df_meses.loc['Enero']

2010    100
2011    120
2012    135
2013    125
2014    245
2015    275
2016    170
2017    220
2018    100
2019    280
2020    150
Name: Enero, dtype: int64

Ahora muestre la facturación de Agosto de 2011 (el 2011 posiblemente sea sin comillas)

In [53]:
# Escriba aquí su código
df_meses.loc['Agosto',2011]

220

Facturaciones de Agosto a Octubre de 2010

In [54]:
# Escriba aquí su código
df_meses.loc['Agosto':"Octubre",2010]

Mes
Agosto        200
Septiembre    170
Octubre       150
Name: 2010, dtype: int64

Facturaciones de Enero a Septiembre de los años 2010 a 2015. 

Nota: posiblemente deba cargar los años con un **range**

In [61]:
# Escriba aquí su código

df_meses.loc['Enero':"Septiembre",range(2010,2015)]

2010  2011  2012  2013  2014
Mes                                     
Enero        100   120   135   125   245
Febrero      120   125   150   180   200
Marzo        135   180   180   125   170
Abril        150   125   250   280   150
Mayo         180   280   245   280   120
Junio        250   280   200   275   105
Julio        245   275   170   220   100
Agosto       200   220   150   200   120
Septiembre   170   200   120   150    25

Como puede ver usando correctamente los índices es posible conseguir un código muy legible. 

### Ejercicio

Cargue el dataset data/breast_cancer.csv, llame *df_breast_cancer* al DataFrame.

In [67]:
# Escriba aquí su código para cargar los datos, 
# cuidado el separador es la línea vertical ...'|' (parece inclinada pero es porque el texto comentado está 
# inclindo, la barra inclinada sería asíp /!!!! Deberá usar sep='|')

df_breast_cancer = pd.read_csv("data/breast-cancer.csv",sep='|')

In [68]:
# Muestre los primeros 7 registros
df_breast_cancer.head(7)

recurrence    age menopause tumor-size inv-nodes node-caps  \
0  no-recurrence-events  30-39   premeno      30-34       0-2        no   
1  no-recurrence-events  40-49   premeno      20-24       0-2        no   
2  no-recurrence-events  40-49   premeno      20-24       0-2        no   
3  no-recurrence-events  60-69      ge40      15-19       0-2        no   
4  no-recurrence-events  40-49   premeno        0-4       0-2        no   
5  no-recurrence-events  60-69      ge40      15-19       0-2        no   
6  no-recurrence-events  50-59   premeno      25-29       0-2        no   

   deg-malig breast breast-quad irradiat  
0          3   left    left_low       no  
1          2  right    right_up       no  
2          2   left    left_low       no  
3          2  right     left_up       no  
4          2  right   right_low       no  
5          2   left    left_low       no  
6          2   left    left_low       no

Veamos un poco de info() acerca del dataFrame

In [69]:
# Escriba aquí su código
df_breast_cancer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286 entries, 0 to 285
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   recurrence   286 non-null    object
 1   age          286 non-null    object
 2   menopause    286 non-null    object
 3   tumor-size   286 non-null    object
 4   inv-nodes    286 non-null    object
 5   node-caps    286 non-null    object
 6   deg-malig    286 non-null    int64 
 7   breast       286 non-null    object
 8   breast-quad  286 non-null    object
 9   irradiat     286 non-null    object
dtypes: int64(1), object(9)
memory usage: 22.5+ KB


Cuántas filas tiene el dataframe?

In [71]:
df_breast_cancer.tail(1)

recurrence    age menopause tumor-size inv-nodes node-caps  \
285  recurrence-events  50-59      ge40      30-34       3-5        no   

     deg-malig breast breast-quad irradiat  
285          3   left    left_low       no

Cuánto nulos hay?

Rta: ninguno

Describa el DataFrame

In [72]:
# Escriba aquí su código
df_breast_cancer.describe()

deg-malig
count  286.000000
mean     2.048951
std      0.738217
min      1.000000
25%      2.000000
50%      2.000000
75%      3.000000
max      3.000000

Cuál es el promedio de deg-malig?

Rta:2.048951 

Cuál es el desvío standard?

Rta:0.738217

Seguramente ha observado que **describe()** sólo nos está mostrando información relevante para las columnas con valores numéricos.  
Sin embargo también puede darnos información útil para columnas cuyos valores sean del tipo string (cadena) o inclusive date (fecha).

Puede ver la documentación en: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html

Probemos: 
**.describe(include = 'all')**

In [73]:
# Escriba aquí su código
df_breast_cancer.describe(include = 'all')

recurrence    age menopause tumor-size inv-nodes node-caps  \
count                    286    286       286        286       286       286   
unique                     2      6         3         11         7         3   
top     no-recurrence-events  50-59   premeno      30-34       0-2        no   
freq                     201     96       150         60       213       222   
mean                     NaN    NaN       NaN        NaN       NaN       NaN   
std                      NaN    NaN       NaN        NaN       NaN       NaN   
min                      NaN    NaN       NaN        NaN       NaN       NaN   
25%                      NaN    NaN       NaN        NaN       NaN       NaN   
50%                      NaN    NaN       NaN        NaN       NaN       NaN   
75%                      NaN    NaN       NaN        NaN       NaN       NaN   
max                      NaN    NaN       NaN        NaN       NaN       NaN   

         deg-malig breast breast-quad irradiat  
count   286.000000    286         286      286  
unique         NaN      2           6        2  
top            NaN   left    left_low       no  
freq           NaN    152         110      218  
mean      2.048951    NaN         NaN      NaN  
std       0.738217    NaN         NaN      NaN  
min       1.000000    NaN         NaN      NaN  
25%       2.000000    NaN         NaN      NaN  
50%       2.000000    NaN         NaN      NaN  
75%       3.000000    NaN         NaN      NaN  
max       3.000000    NaN         NaN      NaN

Fijemonos un poco la información que nos está brindando para las variables de tipo cadena.  

Prestemos atención a la columna **age**:  

Nos dice que:  

- hay 286 valores
- hay 6 valores únicos (es decir distintos)
- el que tiene mayor frecuencia, es decir el que ocurre mayor cantidad de veces es '50-59'
- y ocurre 96 veces.
- El resto de los valores son NaN porque como no es una columna numérica no los puede calcular.


Ahora que sabemos que hay 6 valores distintos para age, podemos estar interesados en saber cuáles son, podemos usar:  

**.unique()**

In [76]:
# Escriba aquí su código
df_breast_cancer.age.unique()

array(['30-39', '40-49', '60-69', '50-59', '70-79', '20-29'], dtype=object)

Cuáles son los distintos valores que puede asumir la columna menopause?

In [77]:
# Escriba aquí su código
df_breast_cancer.menopause.unique()

array(['premeno', 'ge40', 'lt40'], dtype=object)

### EJERCICIO

Muestre los primeros 4 registros del dataframe df_iris

In [78]:
# Escriba aquí su código


sepal length\tsepal width\tpetal length\tpetal width\tiris
0                                      c\tc\tc\tc\td        
1                                      \t\t\t\tclass        
2                    5.1\t3.5\t1.4\t0.2\tIris-setosa        
3                    4.9\t3.0\t1.4\t0.2\tIris-setosa

Qué valores se han cargado en la columna iris?

In [ ]:
# Escriba aquí su código


### EJERCICIO

Ahora trabajemos un poco con el DataFrame df_breast_cancer.  

- Muestre los nombres de los campos o columnas

In [79]:
df_breast_cancer.columns

Index(['recurrence', 'age', 'menopause', 'tumor-size', 'inv-nodes',
       'node-caps', 'deg-malig', 'breast', 'breast-quad', 'irradiat'],
      dtype='object')

- Muestre todos los valores de las filas de índice 11 y 12

In [80]:
# Escriba aquí su código
df_breast_cancer.loc[11:12,:]

recurrence    age menopause tumor-size inv-nodes node-caps  \
11  no-recurrence-events  50-59      ge40      25-29       0-2        no   
12  no-recurrence-events  60-69      lt40      10-14       0-2        no   

    deg-malig breast breast-quad irradiat  
11          2   left    left_low       no  
12          1   left    right_up       no

- Muestre todos los valores de las filas de índice 1, 4 y 10

In [81]:
# Escriba aquí su código
df_breast_cancer.iloc[[1,4,10]]

recurrence    age menopause tumor-size inv-nodes node-caps  \
1   no-recurrence-events  40-49   premeno      20-24       0-2        no   
4   no-recurrence-events  40-49   premeno        0-4       0-2        no   
10  no-recurrence-events  40-49   premeno        0-4       0-2        no   

    deg-malig breast breast-quad irradiat  
1           2  right    right_up       no  
4           2  right   right_low       no  
10          3   left     central       no

- Muestre todas las filas de índice par (es posible que primero deba determinar cuántas filas tiene el df, puede usar shape, pero deberá tener cuidado porque el dataframe tiene filas y columnas)

In [82]:
# Escriba aquí su código para obtener la cantidad de filas
df_breast_cancer.shape[0]

286

In [83]:
# Escriba aquí su código para generar los índices de las filas pares, llame a la variable filas_pares
filas_pares=np.arange(0,df_breast_cancer.shape[0]-1,2) 
filas_pares

array([  0,   2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,
        26,  28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,
        52,  54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,
        78,  80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102,
       104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128,
       130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154,
       156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180,
       182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206,
       208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232,
       234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258,
       260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284])

In [84]:
# Escriba aquí su código para obtener los registros correspondientes a las filas con índice par, muestre sólo los últimos 3
df_breast_cancer.iloc[filas_pares].tail(3)

recurrence    age menopause tumor-size inv-nodes node-caps  \
280  recurrence-events  50-59      ge40      40-44       6-8       yes   
282  recurrence-events  30-39   premeno      20-24       0-2        no   
284  recurrence-events  40-49      ge40      30-34       3-5        no   

     deg-malig breast breast-quad irradiat  
280          3   left    left_low      yes  
282          3   left     left_up      yes  
284          3   left    left_low       no